manual/source/pool/amc.rst

```diff
@@ -139,3 +139,65 @@ AMC interface
             MPS_ARGS_ADD(args, MPS_KEY_FORMAT, fmt);
             res = mps_pool_create_k(&pool, arena, mps_class_amc(), args);
         } MPS_ARGS_END(args);
+
+    When creating an :term:`allocation point` on an AMC pool,
+    :c:func:`mps_ap_create_k` accepts one optional keyword argument:
+
+    * :c:macro:`MPS_KEY_AP_HASH_ARRAYS` (type :c:type:`mps_bool_t`,
+      defaulting to false) specifies (if true) that blocks allocated
+      from the allocation point do not contribute to the *new size* of
+      the :term:`nursery space` for the purposes of deciding whether
+      to start a collection of that generation. See
+      :ref:`pool-amc-hash-arrays`.
+
+
+.. index::
+   pair: AMC pool class; hash arrays
+
+.. _pool-amc-hash-arrays:
+
+Hash arrays
+-----------
+
+the :term:`location dependency` feature of the MPS allows the
+:term:`client program` to implement address-based hash tables in pools
+like AMC that use a :term:`moving memory manager`, re-hashing the
+tables when the addresses they contain might have moved.
+
+However, when a frequently-used hash table grows large enough, the
+following sequence of events may take place:
+
+1. The hash table discovers that its location dependency is stale.
+
+2. A new array is allocated to containt the re-hashed keys.
+
+3. The new array is large enough to push the *new size* of the
+   :term:`nursery space` (that is, the amount of newly allocated
+   memory since the last collection in the first :term:`generation` in
+   the :term:`generation chain` for the pool containing the array)
+   close to its capacity.
+
+4. A small amount of additional allocation causes the new size of the
+   nursery generation to exceed its capacity, which causes the MPS to
+   start a new collection of that generation. This in turn causes the
+   hash table to become stale again.
+
+When the hash table reaches this critical size, the client program may
+find that a large fraction of its time is being spent re-hashing the
+table.
+
+In order to avoid this happening, the MPS provides a mechanism for
+specifying that the newly allocated array does not contribute to the
+new size of the nursery space: this cuts off the vicious cycle at step
+3. To use this mechanism, pass true for the
```

Comment:
comment re paragraph: Fix in d08a1989ada8ff43f5288ee4629dbea7313e73db

comment re "true": We do use "true" and "false" elsewhere in documentation where it's clear that the type concerned is an mps_bool_t. Fix by rewriting this part to avoid reference to "true" in d08a1989ada8ff43f5288ee4629dbea7313e73db